sklearn/feature_extraction/text.py

```diff
@@ -1068,17 +1068,23 @@ def fit(self, X, y=None):
         """
         if not sp.issparse(X):
             X = sp.csc_matrix(X)
+
+        # force convert to np.float*
+        X = check_array(X, accept_sparse='csc', warn_on_dtype=True,
```

Comment:
This will convert X to a float array. Do we really want that.

Reply:
Yes, this was suggested by @rth. Also, if we don't, then the test cases don't comply with int* values. Furthermore, see #8044

Reply:
I've not reviewed the discussion, but I don't get why we are casting X in fit. Can't we just cast df?

Reply:
It's discussed in https://github.com/scikit-learn/scikit-learn/pull/10443#issuecomment-357459029 and in the 3  following comments.

Your opinion on that would we helpful.

Reply:
I still don't get why TfidfTransformer.fit needs to copy X unless X is not CSR/CSC. It should use the default dtype='numeric' in check_array, then use `df = df.astype(X.dtype if X.dtype.kind == 'f' else np.float64)`.

Reply:
> dtype='numeric'

yep, I messed up [there](https://github.com/glemaitre/scikit-learn/commit/4e5e1f06ed1b90c0ffb00584db81a4e8c77e1dff#diff-904f5f849e36c648b8b05de140d7fad8R1076) but it was what I intended with `DTYPES`

Reply:
> df = df.astype(X.dtype if X.dtype.kind == 'f' else np.float64)

we could avoid that by using dtype in log

```python
np.log(n_samples / df, dtype=np.float64) + 1
```